content/lectures/lec05.md

```diff
@@ -3,8 +3,149 @@ title: "Practical Kubernetes"
+
+## Volumes
+
+[Volumes](https://kubernetes.io/docs/concepts/storage/volumes/) are a way to reference some kind of non-ephemeral, persistent storage from within the cluster. They're used to provide filesystem-based persistence across application restarts or node reschedules. There are [PersistentVolumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) (PV) which is a physical representation of a disk in the API. Then we have PersistentVolumeClaims (PVCs) which is a request for access to a PersistentVolume. Then we can bind PVCs to PVs.
+
+A PV is really a representation of the physical block where storage exists, and the PVC is a request to access that storage. This feels a little over-engineered, but it is actually really helpful when using cloud services like AWS where volumes are allocated independently of the machine running the containers.
+
+This is one of the key components of Kubernetes that lets us treat our serves as cattle, not pets. When volumes and storage are separated from our application code, we're able to separate our concerns around running our code from persisting our data. 
+
+## StatefulSet
+
+[StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) is a deployment where each pod gets a persistent volume. This is used for applications that need to be resistant to reboots. Common applications include:
+- Distributed Databases: MongoDB, CockroachDB, Cassandra
+- Games: Minecraft
+- Search: AWS Elasticsearch
+
+![StatefulSet Diagram][/img/lec05/statefulset.png]
+
+Note that each StatefulSet has its own PersistentVolume, and so writes to PersistentVolume 0 from Pod 0 in the diagram will not be readable from either Pod 1 or Pod 2. The application itself will need to coordinate the state of the PersistentVolumes using some kind of replication strategy.
+
+## DaemonSet
+
+A [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/) is a deployment which has one pod running on each node in the cluster. These are used to provide node-wide services like authentication gateways, ingress routers, and log collectors.
+
+![DaemonSet Diagram][/img/lec05/daemonset.png]
+
+## Annotations
+
+[Annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) aren't resources on their own, but a type of resource metadata used to store auxillary key/value information about a resource. Annotations aren't used by Kubernetes itself, but other software in and around your cluster might use the extra metadata to determine its own behavior. 
```

Comment:
Annotations are used by Kubernetes. For example, kubectl stores its last applied config in an annotation. The key thing to realize about labels vs annotations is that annotations are non-identifying i.e. they're only used for storing metadata, not for doing resource lookups.

Reply:
Also an example of an annotation here would be helpful

Reply:
doesn’t kubernetes itself use the status field on resources to handle things, rather than annotations?